File: lib/genetic_algorithm/individuals/EphemeralKeyIndividual.py
import random
from lib.genetic_algorithm.individuals.InputIndividual import InputIndividual
from lib.input_scripts.InputScript import InputScript


class EphemeralKeyIndividual(InputIndividual):
    """
    An individual composed of single key genes and ephemeral key genes
    Can't represent two or more inputs at a time
    """

    def __init__(self, genes):
        InputIndividual.__init__(self, genes)

        self.max_frame = 0
        for gene in genes:
            self.max_frame += gene.get_frames()

    def get_inputs(self):
        """
        Gets the input script of the individual
        :return: <InputScript> the input script that this individual represents
        """
        inputs = InputScript()

        current_frame = 1
        for gene in self.genes:
            gene_input = gene.get_key_input()
            gene_duration = gene.get_frames()
            inputs.add_input(gene_input, current_frame, current_frame + gene_duration)
            current_frame += gene_duration

        return inputs

    def mutate(self):
        """
        Mutates a random gene of the current individual
        """
        random_gene = random.choice(self.genes)
        random_gene.mutate_with_max_frame(random_gene.get_frames + 10)

    def get_max_frame(self):
        """
        getter
        """
        return self.max_frame

File: love_ga_wrapper/love_ga_wrapper/lib/genetic_algorithms/individuals/EphemeralKeyIndividual.lua
local class = require("love_ga_wrapper.lib.classes.class")
local extend = require("love_ga_wrapper.lib.classes.extend")
local InputIndividual = require("love_ga_wrapper.lib.genetic_algorithms.individuals.InputIndividual")
local InputGenomesFactory = require("love_ga_wrapper.lib.genetic_algorithms.genomes.factories.InputGenomesFactory")
-------------------------------------------------------------

-- class: EphemeralKeyIndividual
-- param: genomes:list(SingleInputKeyGenome / EphemeralKeyGenome / NoInputEphemeralKeyGenome) -> The list of genomes of
--                          the current individual
-- An individual with ephemeral genomes and key frame genomes in the genetic algorithm
local EphemeralKeyIndividual = extend(InputIndividual, function(self, genomes)
    self.genome_factory = InputGenomesFactory.new(genomes[1]:getWordDict())
    
    local max_frame = 0
    for i = 1, (# genomes) do
        max_frame = max_frame + genomes[i]:getFrames()
    end
    self.max_frame = max_frame
end)

-- getInputsOnFrame: int(1,) -> dict(str,bool)
-- Gets this individual's inputs given a frame
function EphemeralKeyIndividual.getInputsOnFrame(self, frame)
    local current_frame = 1
    for _, genome in pairs(self.genomes) do
        current_frame = current_frame + genome:getFrames()
        if current_frame > frame then
            return genome:getInputs()
        end
    end
    return {}
end

-- getInputs: None -> list(dict(str, bool))
-- Gets the complete list of this individual's inputs
function EphemeralKeyIndividual.getInputs(self)
    local current_frame = 1
    local inputs = {}
    for i = 1, (# self.genomes) do
        local genome_frames = self.genomes[i]:getFrames()
        local genome_inputs = self.genomes[i]:getInputs()
        inputs[current_frame] = genome_inputs
        inputs[current_frame + genome_frames - 1] = {}
        for input, val in pairs(genome_inputs) do
            inputs[current_frame + genome_frames - 1][input] = not val
        end
        current_frame = current_frame + genome_frames
    end
    return inputs
end

-- getClass: None -> class
-- Returns the class of the Ephemeral Key Individual
function EphemeralKeyIndividual.getClass(self)
  return EphemeralKeyIndividual
end

-- getter
function EphemeralKeyIndividual.getMaxFrame(self)
    return self.max_frame
end

-- mutate: None -> None
-- Mutates a single random gene from the individual
function EphemeralKeyIndividual.mutate(self)
  local random_genome_index  = math.random(1, (# self.genomes))
  self.genomes[random_genome_index]:mutateWithMaxFrame(self:getMaxFrame())
end

return EphemeralKeyIndividual